extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -76,36 +95,27 @@ export class AmpInstallServiceWorker extends AMP.BaseElement {
             'source (%s) or canonical URL (%s) of the AMP-document.',
             origin, sourceUrl.origin, canonicalUrl.origin);
         this.iframeSrc_ = iframeSrc;
-        this.scheduleIframeLoad_();
+        this.insertIframe_();
       }
-      return;
     }
+  }
 
-    if (parseUrl(win.location.href).origin == parseUrl(src).origin) {
-      this.loadPromise(this.win).then(() => {
-        return install(this.win, src);
-      });
-    } else {
-      this.user().error(TAG,
-          'Did not install ServiceWorker because it does not ' +
-          'match the current origin: ' + src);
-    }
+  /** @override */
+  renderOutsideViewport() {
+    // We want the service worker to be installed wherever the element is
+    // located in the document.
+    return true;
   }
 
   /** @override  */
   getPriority() {
     return 3;
   }
 
-  /** @private */
-  scheduleIframeLoad_() {
-    Services.viewerForDoc(this.getAmpDoc()).whenFirstVisible().then(() => {
-      this.deferMutate(this.insertIframe_.bind(this));
-    });
-  }
-
   /** @private */
   insertIframe_() {
+    // Mutate block is not required when calling this method, since the inserted
+    // iframe has `display: none` and does not modify the document's layout.
```

Comment:
Unfortunately not, any insertion can cause CSS style recalc.

Reply:
Is it enough to call it from inside layoutCallback() or does it need more explicit protection?